### PR TITLE
Artifact replication survives peer restart

### PR DESCRIPTION
**Pull request by gd-tmagrys**
https://github.com/griddynamics/nexus-replication-plugin/issues/12

Please review and merge.

### CODE REVIEW

**Comment by ctapobep** [comment]
This PR depends on a commit from another PR. Is that expected?

**Comment by gd-tmagrys** [comment]
@ctapobep answering your question "This PR depends on a commit from another PR. Is that expected?", my answer is Yes. This was done by purpose. Once I get that PR merged, I'll rebase on master, and recreate PR if needed.

File: src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java [hunk 1]
```diff
@@ -28,16 +28,25 @@
...
 public class NexusServer {
-    @Getter
     @NonNull
     @XmlElement(name = "url")
     private String url;
-    @Getter
     @NonNull
     @XmlElement(name = "user")
     private String user;
-    @Getter
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java]: What's the purpose of replacing `@Getter` with getter method?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 2]
```diff
@@ -50,23 +54,83 @@
...
-                30, TimeUnit.SECONDS, queue);
+                30,
+                TimeUnit.SECONDS,
+                new LinkedBlockingQueue<Runnable>(
+                        configurationsManager.getConfiguration().getRequestsQueueSize())
+        );
+        initBackgroundWorkers(configurationsManager);
+    }
+
+    private void initBackgroundWorkers(ConfigurationsManager configurationsManager) {
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Just pass `ReplicationPluginConfiguration` - this method doesn't need more.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 3]
```diff
@@ -50,23 +54,83 @@
...
+                        } catch (Exception e) {
+                            log.error(e.getMessage(), e);
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    private FileBlockingQueue initFileBlockingQueue(ConfigurationsManager configurationsManager) {
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Just pass `ReplicationPluginConfiguration` - this method doesn't need more.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 4]
```diff
@@ -50,23 +54,83 @@
...
+        BlockingQueue<ArtifactMetaInfo> blockingQueue =
+                new LinkedBlockingQueue<>(configurationsManager.
+                        getConfiguration().getRequestsQueueSize());
+        String blockingQueueDumpFileName = configurationsManager.getConfiguration().getQueueDumpFileName();
+        FileBlockingQueue retVal = new FileBlockingQueue(blockingQueue,
+                blockingQueueDumpFileName);
+        try {
+            File blockingQueueInputFile = new File(blockingQueueDumpFileName);
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: It's a `queueInputFile`, the file doesn't block..

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 5]
```diff
@@ -50,23 +54,83 @@
...
+        int requestsSendingThreadsCount = configurationsManager.getConfiguration()
+                .getRequestsSendingThreadsCount();
+        ExecutorService executorService = Executors.newFixedThreadPool(requestsSendingThreadsCount);
+        for (int i = 0; i < requestsSendingThreadsCount; i++) {
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        try {
+                            final ArtifactMetaInfo artifactMetaInfo = fileBlockingQueue.peek();
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Does it meed to be final?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 6]
```diff
@@ -50,23 +54,83 @@
...
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+        return retVal;
+    }
+
+    @Override
+    public void offerRequest(ArtifactMetaInfo artifactMetaInfo) {
+        try {
+            fileBlockingQueue.offer(artifactMetaInfo, 30, TimeUnit.SECONDS);
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Could you please extract 30 into a constant and add JavaDocs explaining why it's set to 30?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 7]
```diff
@@ -119,14 +185,15 @@ public void onComplete(Future<RestResponse> future) throws InterruptedException
...
     /**
      * Creates jersey HTTP client
-     * @param login Username on the remote server
+     *
+     * @param login    Username on the remote server
      * @param password User's password
      * @return HTTP client
      */
     private Client getClient(String login, String password) {
         ClientConfig config = new DefaultClientConfig();
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: Looks like this config doesn't have anything configured.. What we need to set is _relatively_ low timeouts for connection creation and for connection duration. Just to illustrate what may happen:
- Nexus Receiver is behind a firewall and someone just configured FW rules in a wrong way. FW silently drops packages from Nexus Sender.
- Nexus Sender has large timeouts and therefore waits for 30 mins on every HTTP request to the Receiver.

Result: all sending threads are blocked for a _very_ long time and there is no way to fix that without Nexus Sender restart.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java [hunk 8]
```diff
@@ -34,6 +35,9 @@
 import javax.inject.Singleton;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.UriBuilder;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.Unmarshaller;
+import java.io.File;
 import java.util.concurrent.*;
 
 @Singleton
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java]: This class needs JavaDocs:
```
Responsible to send request to other Nexus instances to notify them about new artifacts. There are 2 separate thread pools working along:
- Usual sending threads that work when new artifacts are uploaded to Nexus Sender
- Threads that read queue of artifacts from file. That file is filled with new artifacts if Nexus Receiver was not available and our Nexus was shut down. It then reads artifacts from the file and tries to send them to Receiver again.
```

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.Marshaller;
+import java.io.File;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+public class FileBlockingQueue {
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java]:
```
Stores artifacts that Plugin failed to send to another Nexus instance. If that happens, we don't want
to loose the history of updates and therefore we'll retry sending the requests even if Nexus Sender
was restarted.
```

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }
+
+    private synchronized void saveQueueToFile() {
+
+        try {
+            String backupBlockingQueueDumpFileName = blockingQueueDumpFileName + ".bak";
+            File blockingQueueDumpFile = new File(blockingQueueDumpFileName);
+            if (blockingQueueDumpFile.exists() && !blockingQueueDumpFile.isDirectory()) {
+                FileUtils.copyFile(blockingQueueDumpFile, new File(backupBlockingQueueDumpFileName));
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java]: Still seems weird to me, but since you're saying this may be replaced with embedded DB, I'm okay. Could you please raise an issue so that we don't forget to do that?

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.File;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+public class FileBlockingQueue {
+
+    private final BlockingQueue<ArtifactMetaInfo> internalBlockingQueue;
+    private final String blockingQueueDumpFileName;
+
+    private Logger log = LoggerFactory.getLogger(FileBlockingQueue.class);
```
Comment (ctapobep) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java]: `final`

**Comment by ctapobep** [comment]
A general note - please pay attention to variable names and to formatting. It often makes it harder to read. And don't forget about JavaDocs :)

File: src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java (hunk 1 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java]: The only reason:
Make IntelliJ IDEA do not underline missing method with red.

Comment (alekseysemenov) [inline src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java]: You could use this plugin https://plugins.jetbrains.com/plugin/6317 to resolve "missing" code.

File: src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java (hunk 10 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/FileBlockingQueue.java]: Done:
https://github.com/griddynamics/nexus-replication-plugin/issues/31

File: src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java (hunk 1 above)
Comment (gd-tmagrys) [inline src/main/java/com/griddynamics/cd/nrp/internal/model/config/NexusServer.java]: Thanks for this link. I revert this change back to use @Getter

**Comment by gd-tmagrys** [comment]
Seems that I fixed all remarks.

**Comment by ctapobep** [comment]
Let's merge then. 
FYI: I didn't test the plugin, just reviewed the code. So hopefully it's all tested :)

**Comment by gd-tmagrys** [comment]
I was testing it :+1:

**Comment by kevstigneev** [comment]
@alekseysemenov please merge it.
